File: gilded_rose/src/main/java/gildedrose/GildedRose.java
package gildedrose;
import java.util.ArrayList;
import java.util.List;


public class GildedRose {

	public Item items [];

    public GildedRose(Item[] items) {
        this.items = items;
    }


    /**
	 * @param args
	 */
	public void main(String[] args) {
		
        System.out.println("OMGHAI!");

        items = new Item[]{
                new Item("+5 Dexterity Vest", 10, 20),
        new Item("Aged Brie", 2, 0),
        new Item("Elixir of the Mongoose", 5, 7),
        new Item("Sulfuras, Hand of Ragnaros", 0, 80),
        new Item("Backstage passes to a TAFKAL80ETC concert", 15, 20),
        new Item("Conjured Mana Cake", 3, 6)};
        updateQuality();
}


	
    public void updateQuality()
    {
        for (Item item : items) {
            ItemUpdatable itemUpdatable = (ItemUpdatable) item;
            itemUpdatable.doUpdateQuality();
        }
    }

}

File: gilded_rose-scala/src/main/scala/com/gildedrose/GildedRose.scala
package com.gildedrose

import com.gildedrose.GildedRose.{AGEDBRIE, BACKSTAGE, SULFURAS}


class GildedRose(val items: Array[Item]) {
  def updateQuality() {
    items.foreach { item =>
      item.name match {
        case AGEDBRIE =>
          updateAgedBrie(item)
        case BACKSTAGE =>
          updateBackstage(item)
        case SULFURAS =>
          updateSulfuras(item)
        case _ =>
          updateDefault(item)
      }
    }
  }

  private def updateDefault(item: Item) = {
    if (item.quality > 0) {
      item.quality = item.quality - 1
    }

    item.sellIn = item.sellIn - 1

    if (item.sellIn < 0) {
      if (item.quality > 0) {
        item.quality = item.quality - 1
      }
    }
  }

  private def updateSulfuras(item: Item) = {

  }

  private def updateBackstage(item: Item) = {
    if (item.quality < 50) {
      item.quality = item.quality + 1
      if (item.sellIn < 11) {
        if (item.quality < 50) {
          item.quality = item.quality + 1
        }
      }
      if (item.sellIn < 6) {
        if (item.quality < 50) {
          item.quality = item.quality + 1
        }
      }
    }
    item.sellIn = item.sellIn - 1
    if (item.sellIn < 0) {
      item.quality = 0
    }
  }

  private def updateAgedBrie(item: Item) {
    if (item.quality < 50) {
      item.quality = item.quality + 1
    }
    item.sellIn = item.sellIn - 1
    if (item.sellIn < 0) {
      if (item.quality < 50) {
        item.quality = item.quality + 1
      }
    }
  }

}

object GildedRose {
  private val AGEDBRIE = "Aged Brie"
  private val SULFURAS = "Sulfuras, Hand of Ragnaros"
  private val BACKSTAGE = "Backstage passes to a TAFKAL80ETC concert"
}